rpc/lib/client/ws_client.go

```diff
@@ -1,160 +1,390 @@
+	mtx sync.RWMutex
+
+	// Time allowed to read the next pong message from the server.
+	pongWait time.Duration
+
+	// Send pings to server with this period. Must be less than pongWait.
+	pingPeriod time.Duration
 }
 
-// create a new connection
-func NewWSClient(remoteAddr, endpoint string) *WSClient {
+// NewWSClient returns a new client.
+func NewWSClient(remoteAddr, endpoint string, options ...func(*WSClient)) *WSClient {
 	addr, dialer := makeHTTPDialer(remoteAddr)
-	wsClient := &WSClient{
-		Address:  addr,
-		Dialer:   dialer,
-		Endpoint: endpoint,
-		Conn:     nil,
+	c := &WSClient{
+		Address:              addr,
+		Dialer:               dialer,
+		Endpoint:             endpoint,
+		PingPongLatencyTimer: metrics.NewTimer(),
+		pongWait:             defaultPongWait,
+		pingPeriod:           defaultPingPeriod,
+	}
+	c.BaseService = *cmn.NewBaseService(nil, "WSClient", c)
+	for _, option := range options {
+		option(c)
+	}
+	return c
+}
+
+// PingPong allows changing ping period and pong wait time. If ping period
+// greater or equal to pong wait time, panic will be thrown.
+func PingPong(pingPeriod, pongWait time.Duration) func(*WSClient) {
+	return func(c *WSClient) {
+		if pingPeriod >= pongWait {
+			panic(fmt.Sprintf("ping period (%v) must be less than pong wait time (%v)", pingPeriod, pongWait))
+		}
+		c.pingPeriod = pingPeriod
+		c.pongWait = pongWait
 	}
-	wsClient.BaseService = *cmn.NewBaseService(nil, "WSClient", wsClient)
-	return wsClient
 }
 
-func (wsc *WSClient) String() string {
-	return wsc.Address + ", " + wsc.Endpoint
+// String returns WS client full address.
+func (c *WSClient) String() string {
+	return fmt.Sprintf("%s (%s)", c.Address, c.Endpoint)
 }
 
-// OnStart implements cmn.BaseService interface
-func (wsc *WSClient) OnStart() error {
-	wsc.BaseService.OnStart()
-	err := wsc.dial()
+// OnStart implements cmn.Service by dialing a server and creating read and
+// write routines.
+func (c *WSClient) OnStart() error {
+	err := c.dial()
 	if err != nil {
 		return err
 	}
-	wsc.ResultsCh = make(chan json.RawMessage, wsResultsChannelCapacity)
-	wsc.ErrorsCh = make(chan error, wsErrorsChannelCapacity)
-	go wsc.receiveEventsRoutine()
+
+	c.ResultsCh = make(chan json.RawMessage)
+	c.ErrorsCh = make(chan error)
+
+	c.send = make(chan types.RPCRequest)
+	// 1 additional error may come from the read/write
+	// goroutine depending on which failed first.
+	c.reconnectAfter = make(chan error, 1)
+	// capacity for 1 request. a user won't be able to send more because the send
+	// channel is unbuffered.
+	c.backlog = make(chan types.RPCRequest, 1)
+
+	c.startReadWriteRoutines()
+	go c.reconnectRoutine()
+
 	return nil
 }
 
-// OnReset implements cmn.BaseService interface
-func (wsc *WSClient) OnReset() error {
-	return nil
+// OnStop implements cmn.Service.
+func (c *WSClient) OnStop() {}
+
+// Stop overrides cmn.Service#Stop. There is no other way to wait until Quit
+// channel is closed.
+func (c *WSClient) Stop() bool {
```

Comment:
I want to close ResultsCh and ErrorsCh only when we can't write to them (to avoid panic). There is no other way to wait until Quit channel is closed other than overriding Stop.